src/v/cloud_storage/segment_chunk_api.cc

```diff
@@ -0,0 +1,344 @@
+    return ss::now();
+}
+
+ss::future<> segment_chunks::stop() {
+    vlog(_ctxlog.debug, "stopping segment_chunks");
+    _eviction_timer.cancel();
+    if (!_as.abort_requested()) {
+        _as.request_abort();
+    }
+
+    co_await _gate.close();
+    vlog(_ctxlog.debug, "stopped segment_chunks");
+}
+
+bool segment_chunks::downloads_in_progress() const {
+    return std::any_of(_chunks.begin(), _chunks.end(), [](const auto& entry) {
+        return entry.second.current_state == chunk_state::download_in_progress;
+    });
+}
+
+ss::future<ss::file>
+segment_chunks::do_hydrate_and_materialize(chunk_start_offset_t chunk_start) {
+    gate_guard g{_gate};
+    vassert(_started, "chunk API is not started");
+
+    auto it = _chunks.find(chunk_start);
+    std::optional<chunk_start_offset_t> chunk_end = std::nullopt;
+    if (auto next = std::next(it); next != _chunks.end()) {
+        chunk_end = next->first - 1;
+    }
+
+    co_await _segment.hydrate_chunk(chunk_start, chunk_end);
+    co_return co_await _segment.materialize_chunk(chunk_start);
+}
+
+ss::future<segment_chunk::handle_t>
+segment_chunks::hydrate_chunk(chunk_start_offset_t chunk_start) {
+    gate_guard g{_gate};
+    vassert(_started, "chunk API is not started");
+
+    vassert(
+      _chunks.contains(chunk_start),
+      "No chunk starting at offset {}, cannot hydrate",
+      chunk_start);
+
+    auto& chunk = _chunks[chunk_start];
+    auto curr_state = chunk.current_state;
+    if (curr_state == chunk_state::hydrated) {
+        vassert(
+          chunk.handle,
+          "chunk state is hydrated without data file for id {}",
+          chunk_start);
+        co_return chunk.handle.value();
+    }
+
+    // If a download is already in progress, subsequent callers to hydrate are
+    // added to a wait list, and notified when the download finishes.
+    if (curr_state == chunk_state::download_in_progress) {
+        co_return co_await add_waiter_to_chunk(chunk_start, chunk);
+    }
+
+    // Download is not in progress. Set the flag and begin download attempt.
+    try {
+        chunk.current_state = chunk_state::download_in_progress;
+
+        // Keep retrying if materialization fails.
+        bool done = false;
+        while (!done) {
+            auto handle = co_await do_hydrate_and_materialize(chunk_start);
+            if (handle) {
+                done = true;
+                chunk.handle = ss::make_lw_shared(std::move(handle));
+            } else {
+                vlog(
+                  _ctxlog.trace,
+                  "do_hydrate_and_materialize failed for chunk start offset {}",
+                  chunk_start);
+                co_await ss::sleep_abortable(
+                  _cache_backoff_jitter.next_jitter_duration(), _as);
+            }
+        }
+    } catch (const std::exception& ex) {
+        while (!chunk.waiters.empty()) {
+            chunk.waiters.front().set_to_current_exception();
+            chunk.waiters.pop_front();
+        }
```

Comment:
should we set `chunk.current_state` back to `not_available` here?

Reply:
good catch, fixed